webpack/assets/javascripts/react_app/components/Layout/LayoutActions.js

```diff
@@ -0,0 +1,181 @@
+import {
+  LAYOUT_SHOW_LOADING,
+  LAYOUT_HIDE_LOADING,
+  LAYOUT_CHANGE_ACTIVE,
+  LAYOUT_RESOURCES_REQUEST,
+  LAYOUT_CHANGE_LOCATION,
+  LAYOUT_CHANGE_ORG,
+} from './LayoutConstants';
+
+export const showLoading = () => ({
+  type: LAYOUT_SHOW_LOADING,
+});
+
+export const hideLoading = () => ({
+  type: LAYOUT_HIDE_LOADING,
+});
+
+export const changeActiveMenu = primary => (dispatch) => {
+  dispatch({
+    type: LAYOUT_CHANGE_ACTIVE,
+    payload: {
+      primary,
+    },
+  });
+};
+
+export const fetchMenuItems = menuItems => (dispatch) => {
+  const activePath = window.location.pathname;
+  const items = customItems(menuItems, activePath);
+
+  dispatch({
+    type: LAYOUT_RESOURCES_REQUEST,
+    payload: {
+      items,
+    },
+  });
+};
+
+export const changeOrg = (id, org) => (dispatch) => {
+  // use ID for api
+  dispatch({
+    type: LAYOUT_CHANGE_ORG,
+    payload: {
+      org,
+    },
+  });
+};
+
+export const changeLoc = (id, location) => (dispatch) => {
+  // use ID for api
+  dispatch({
+    type: LAYOUT_CHANGE_LOCATION,
+    payload: {
+      location,
+    },
+  });
+};
+
+const customItems = (data, activePath) => {
+  const items = [];
+
+  // Menu Items
+
+  data.menu.forEach((menu) => {
+    menu.forEach((item) => {
+      let activeFlag = false;
+      const childrenArray = [];
+      item.children.forEach((child) => {
+        if (child.url === activePath) activeFlag = true;
```

Comment:
we can gain much performance and cleaner code if this is extracted into selectors.
you can use [filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) instead forEach.

Reply:
Why filter? i don't want to remove any of the elements, only change their active status